parsing/location.ml

```diff
@@ -377,172 +385,210 @@ let highlight_dumb ~print_chars ppf lb locs =
     Format.fprintf ppf "@]@,@]"
   end
 
-let show_code_at_location ppf lb locs =
-  try highlight_dumb ~print_chars:false ppf lb locs
+let highlight_dumb ~print_chars lb ppf locs =
+  try highlight_dumb ~print_chars lb ppf locs
   with Exit -> ()
 
-(* Highlight the location using one of the supported modes. *)
-
-let rec highlight_locations ppf locs =
-  match !status with
-    Terminfo.Uninitialised ->
-      status := Terminfo.setup stdout; highlight_locations ppf locs
-  | Terminfo.Bad_term ->
-      begin match !input_lexbuf with
-        None -> false
-      | Some lb ->
-          let norepeat =
-            try Sys.getenv "TERM" = "norepeat" with Not_found -> false in
-          if norepeat then false else
-            try highlight_dumb ~print_chars:true ppf lb locs; true
-            with Exit -> false
-      end
-  | Terminfo.Good_term ->
-      begin match !input_lexbuf with
-        None -> false
-      | Some lb ->
-          try highlight_terminfo ppf lb locs; true
-          with Exit -> false
-      end
-
-let default_printer ppf loc =
-  setup_colors ();
-  if loc.loc_start.pos_fname = "//toplevel//"
-  && highlight_locations ppf [loc] then ()
-  else Format.fprintf ppf "@{<loc>%a@}:@," print_loc loc
-;;
-
-let printer = ref default_printer
-let print ppf loc = !printer ppf loc
-
-let print_compact ppf loc =
-  if loc.loc_start.pos_fname = "//toplevel//"
-  && highlight_locations ppf [loc] then ()
-  else begin
-    let (file, line, startchar) = get_pos_info loc.loc_start in
-    let endchar = loc.loc_end.pos_cnum - loc.loc_start.pos_cnum + startchar in
-    Format.fprintf ppf "%a:%i" print_filename file line;
-    if startchar >= 0 then Format.fprintf ppf ",%i--%i" startchar endchar
-  end
-;;
+(* Return the "best" highlighting function depending on the capabilities of the
+   terminal. *)
+let highlight_locations ppf locs =
+  setup_terminal ();
+  let norepeat =
+    try Sys.getenv "TERM" = "norepeat" with Not_found -> false
```

Comment:
Somewhat unrelated to your PR but I noticed there were no tests exercising this `norepeat` behavior.
Could you add one as a preliminary commit? (I assume we can pass some environment variables through ocamltest, is that correct @shindere?)

Reply:
Yes, `flags = "..."` works as expected for the `toplevel` action.

Reply:
That's not what I want though, `flags` is for cmdline flags. Here the behavior is controlled through an environment variable (`TERM`).

Reply:
Good idea, I'll investigate how to test this. It would also be interesting to know what is the motivation behind `TERM=norepeat` -- is it actually used somewhere? Maybe @xavierleroy knows?

Reply:
Digging up a bit, `TERM=norepeat` has been introduced by @garrigue as `TERM=dumb` in b92bf2, then renamed to `character` in 9a3267, then to `norepeat`in 59d5d1. The text of the first commit seems to imply that it was used by the manual, but this is not the case anymore AFAIK.

Reply:
Indeed, relying on `norepeat` in the manual build was removed in 4be6cafcc3b9ac9d5e68e5dfbcc69074e3f4f30e. Feel free to remove that special support in a later PR or this one.

Reply:
I'll do that in a separate PR just in case someone actually uses that in the wild.

Reply:
In theory, `norepeat` is useful when communicating with a toplevel launched in a separate process, where repeating the faulty line is not a very useful behavior. (At least, this was the use case of the manual). This kind of use is partially superseded by compiler-libs , but I am not sure if it is still used in the wild.